lib/db_connection/holder.ex

```diff
@@ -56,35 +56,46 @@ defmodule DBConnection.Holder do
     now = System.monotonic_time(@time_unit)
     timeout = abs_timeout(now, opts)
 
-    case checkout(pool, callers, queue?, now, timeout) do
-      {:ok, _, _, _, _} = ok ->
-        ok
-
-      {:error, _} = error ->
-        error
-
-      {:redirect, caller, proxy} ->
-        case checkout(proxy, opts) do
-          {:ok, _, _, _, _} = ok ->
-            ok
-
-          {:error, %DBConnection.ConnectionError{message: message} = exception} ->
-            {:error,
-             %{
-               exception
-               | message:
-                   "could not checkout the connection owned by #{inspect(caller)}. " <>
-                     "When using the sandbox, connections are shared, so this may imply " <>
-                     "another process is using a connection. Reason: #{message}"
-             }}
-
-          {:error, _} = error ->
-            error
-        end
+    result =
+      case checkout(pool, callers, queue?, now, timeout) do
+        {:ok, _, _, _, _} = ok ->
+          ok
+
+        {:error, _} = error ->
+          error
+
+        {:redirect, caller, proxy} ->
+          case checkout(proxy, opts) do
+            {:ok, _, _, _, _} = ok ->
+              ok
+
+            {:error, %DBConnection.ConnectionError{message: message} = exception} ->
+              {:error,
+               %{
+                 exception
+                 | message:
+                     "could not checkout the connection owned by #{inspect(caller)}. " <>
+                       "When using the sandbox, connections are shared, so this may imply " <>
+                       "another process is using a connection. Reason: #{message}"
+               }}
+
+            {:error, _} = error ->
+              error
+          end
+
+        {:exit, reason} ->
+          exit({reason, {__MODULE__, :checkout, [pool, opts]}})
+      end
 
-      {:exit, reason} ->
-        exit({reason, {__MODULE__, :checkout, [pool, opts]}})
+    case result do
+      {:error, %DBConnection.ConnectionError{} = err} ->
+        Enum.each(opts[:connection_listeners] || [], &send(&1, {:checkout_timeout, err}))
```

Comment:
Those errors are going to be other things besides checkout_timeout, no?

Reply:
I think if you want only the checkout timeouts, then we need to put it close to where the error is actually raised?

Reply:
Yes. Maybe we can relay them all but change the message to a more relevant name?
Checkout timeouts have `reason: :queue_timeout` in the struct

Reply:
> I think if you want only the checkout timeouts, then we need to put it close to where the error is actually raised?

I'm not sure which is better. I can see use for having access to all the errors, but I'm almost sure that the other errors can be inferred through other Ecto Telemetry events... What do you think?

Also, I'd need some help in putting this closer to the error. Another option is having a more strict match in this `case` for identifying the relevant `:reason`

Reply:
Thinking as a maintainer, maybe relaying only `:queue_timeout`s would be better in the sense that adding features/events is easy, but removing them could be problematic

Reply:
> Thinking as a maintainer, maybe relaying only :queue_timeouts would be better in the sense that adding features/events is easy, but removing them could be problematic

Exactly. Although I am still wondering, why not emit this from the connection pool, which is where the error is raised?

Reply:
I couldn't my way well enough through the project's structure. In the exact place where the error is raised the connection_listeners info isn't available (lib/db_connection/connection_pool.ex:293 and lib/db_connection/ownership/proxy.ex:291), so I'd need to be a little higher in the callstack.

Could you point me towards the right direction?

Reply:
btw, I've already pushed the version which only relays `:queue_timeout`